### PR TITLE
Feature/Packet capture using tshark processing of tcpdump capture

### LINKED ISSUE #942: [TASK] Capture packets in test_flows.py using tshark
## Description

Currently, both `test_flows` implementations search the logs in order to confirm that messages were received properly.
However, this causes the logs to be flooded with unnecessary prints and doesn't provide good debugging information for the script, as sometimes it can find irrelevant prints, and the script's string expressions being searched for in the logs need to be identical to the cpp implementation, causing even more confusion.

A proposed improvement to this would be to analyze the raw packet data using the python script.
An attempt was made in PR #924, utilizing tshark to generate a json file from the tcpdump and later opening it in python, however, as suggested by @arnout and @tomereli, a more elegant solution would be to capture the packets directly using tshark, and handling them later as we desire, e.g. filter them or dump them into a file.

## Challenge

At the moment I'm attempting to implement this using [pyshark](https://github.com/KimiNewt/pyshark) however, after attempting to capture the packets in the background on a different thread, I ran into a bug with pyshark's implementation causing an issue when attempting to [run the sniffer in a different thread.](https://github.com/KimiNewt/pyshark/issues/353)
 
So we need to think of a different approach.

### PR DESCRIPTION
**Pull request by Galvien**
As a third attempt, this change allows easy, **fast** and simple analysis of CMDU messages passing through the container's bridge interface.

The general idea is that each running test is monitored by tcpdump, then saves the packet data into a pcap file, which is later quickly scanned by tshark.

In addition, a helper method extends the dictionary with direct, casted entries of more important values from the tree - for easier access and handling.

### CODE REVIEW

**Comment by Galvien** [comment]
@arnout Gitlab CI will always fail since this change completely removes the `--tcpdump` option.
I think we'll need to change that.

**Comment by tomereli** [comment]
> @arnout Gitlab CI will always fail since this change completely removes the `--tcpdump` option.
> I think we'll need to change that.

So change what you need in .gitlab-ci.yml

File: tests/opts.py [hunk 1]
```diff
@@ -9,7 +9,6 @@
 class opts:
     '''Static class that encodes the global options.'''
     verbose = False
```
Comment (tomereli) [inline tests/opts.py:11]: If you remove this section, you need to also update the README to mention that tcpdump is now a requirement for running tests even in docker. You should also explain how to fix the capabilities so that non sudo users can use tcpdump.

File: tests/sniffer.py [hunk 2]
```diff
@@ -13,30 +13,33 @@
...
-                    # Make sure it doesn't block due to stderr buffering
-                    self.tcpdump_proc.stderr.close()
-                    break
-            else:
-                err("tcpdump terminated")
-                self.tcpdump_proc = None
+    def start(self, test_name):
+        '''Start tcpdump.'''
+        outputfile_basename = "test_" + test_name
+        debug("Starting tcpdump, output file {}.pcap".format(outputfile_basename))
```
Comment (tomereli) [inline tests/sniffer.py:106]: I don't think you should make the output file hard coded - why is this needed? You can just add another argument with default value (None) for the output file name - so that if it is None, you will use the test name.

File: tests/sniffer.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ def start(self, test_name):
         os.makedirs(os.path.join(opts.tcpdump_dir, 'logs'), exist_ok=True)
         outputfile = os.path.join(opts.tcpdump_dir, outputfile_basename) + ".pcap"
         self.outputfiles[test_name] = outputfile
-        command = ["tcpdump", "-i", self.interface, '-w', outputfile]
+        command = ["tcpdump", "-i", self.interface,'-U', '-w', outputfile]
```
Comment (tomereli) [inline tests/sniffer.py]: I don't understand whNot sure I understand that the `-U` flag is used for here:

From tcpdump manual:

>-U
--packet-buffered
If the -w option is not specified, or if it is specified but the --print flag is also specified, make the printed packet output ``packet-buffered''; i.e., as the description of the contents of each packet is printed, it will be written to the standard output, rather than, when not writing to a terminal, being written only when the output buffer fills.
If the -w option is specified, make the saved raw packet output ``packet-buffered''; i.e., as each packet is saved, it will be written to the output file, rather than being written only when the output buffer fills.
The -U flag will not be supported if tcpdump was built with an older version of libpcap that lacks the pcap_dump_flush(3PCAP) function.

Your commit message also doesn't explain what issue this flag is solving. Please elaborate.

File: tests/sniffer.py [hunk 4]
```diff
@@ -41,6 +42,23 @@ def start(self, test_name):
...
 
+    def get_packet_capture(self, test_name):
+        if test_name in self.captures:
+            return self.captures[test_name]
+        tshark_command = ['tshark', '-r', self.outputfiles[test_name], '-T', 'json']
+        self.tshark_proc = subprocess.run(
+            tshark_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        try:
+            capture = json.loads(self.tshark_proc.stdout)
+            def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
```
Comment (tomereli) [inline tests/sniffer.py]: I don't think this is appropriate - instead, we should capture only ieee1905 packets from the start when we use tcpdump:
```
tcpdump -i<iface> ether proto 0x893a
```

File: tests/sniffer.py [hunk 5]
```diff
@@ -41,6 +42,23 @@ def start(self, test_name):
...
+            result = self._simplify_packets(filter(ieee_filter, capture))
+            if not result or len(result) == 0:
+                result = None
+            self.captures[test_name] = result
+            return self.captures[test_name]
+        except BaseException as ex:
+            return None
```
Comment (tomereli) [inline tests/sniffer.py]: Can you explain what kind of exception we're ignoring here?

File: tests/sniffer.py [hunk 6]
```diff
@@ -59,6 +59,20 @@ def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
         except BaseException as ex:
             return None
 
+    def _simplify_packets(self, packets):
```
Comment (tomereli) [inline tests/sniffer.py]: This should be reordered since you use this method in the previous commit.

File: tests/test_flows.py [hunk 7]
```diff
@@ -365,10 +365,12 @@ def test_ap_capability_query(self):
 
         debug("Confirming ap capability query has been received on agent")
         self.check_log(env.agents[0], "AP_CAPABILITY_QUERY_MESSAGE")
-
+        capability_query = env.wired_sniffer.get_packets_for_cmdu_type(self.running,0x8001)
```
Comment (tomereli) [inline tests/test_flows.py]: The example should replace the check_log, not just add a call without testing it. Also I think you should add an API to search for specific packets like it is done in the UCC.

For example, this is how the UCC is checking for a specific topology response:

```
2020-04-14 04:08:41.506 - INFO - 
        ~~~~~ 4.1.1 Step 9 sniffer checks - Topology Response message ~~~~~

2020-04-14 04:08:41.512 - INFO - SNIFFER (192.168.250.5:9999) ---> sniffer_control_filter_capture,InFile,_MAP-4.1.1-STEP9,OutFile,MAP-4.1.1-STEP9,msgId,0x43c6,filterfile,mapfile,msgType,0x0003,nframes,1
2020-04-14 04:08:42.023 - INFO - SNIFFER (192.168.250.5:9999) <-- status,COMPLETE,FilterStatus,SUCCESS,FrameCount,1

```

I think the equivalent in python should be something like:

```
topology_response = env.wired_sniffer.get_packets(self.running, cmdu_type=0x0003, mid=0x43c6)
```
The API should also allow for more arguments - source and destination MACs so we can filter on ALid, and probably more that I can't think of now. Every parameter should be optional except for self.running.

File: tests/sniffer.py [hunk 8]
```diff
@@ -73,6 +74,23 @@ def _simplify(x):
             return x
         return list(map(_simplify, packets))
 
+    def get_packets_for_cmdu_type(self, test_name, msg_type: int):
```
Comment (tomereli) [inline tests/sniffer.py]: I think we're missing an important argument here - message identifier (mid) - this one should be used to verify that we are looking at the correct cmdu. This should be optional parameter, so if its none, it should return all the CMDUs which match the message type, otherwise it would return a single CMDU or None if there is no message with this MID.

In addition, we must also add source and destination macs parameters, since this is the only way we can uniquely identify a message - source + destination ALid + mid.

**Review by tomereli — CHANGES_REQUESTED** [review]
Very nice PR, good commits split. I have some comments so re-review is required.

File: tests/sniffer.py [hunk 9]
```diff
@@ -7,36 +7,89 @@
...
+        return list(map(_simplify, packets))
+
+    def get_packets_for_cmdu_type(self, test_name, msg_type: int):
+        def filter_func(x): return x['msg_type'] == msg_type
+        try:
+            return list(filter(filter_func, self.get_packet_capture(test_name)))
+        except:
+            err("No packets captured for specified cmdu type: {}".format(msg_type))
+            return None
+        return list(filter(filter_func, self.get_packet_capture(test_name)))
```
Comment (ghost) [inline tests/sniffer.py]: I think you must not be familiar with list comprehensions in python. They're really useful. 
```python
def get_packets_for_cmdu_type(self, test_name :str, msg_type: int):
    return [x for x in self.get_packet_capture(test_name) if x['msg_type'] == msg_type]
```

File: tests/sniffer.py [hunk 10]
```diff
@@ -7,36 +7,89 @@
...
-            else:
-                err("tcpdump terminated")
-                self.tcpdump_proc = None
+    def filter_packets_by_time(self, packets, start_time=None, end_time=time()):
+        def _get_time_epoch_for_packet(packet):
+            t = packet['time_epoch']
+            if start_time:
+                return t >= start_time and t <= end_time
+            return t <= end_time
+        return list(filter(_get_time_epoch_for_packet, packets))
```
Comment (ghost) [inline tests/sniffer.py]: Same here, I would do this with a list comprehension
```python
def filter_packets_by_time(self, packets, start_time=None, end_time=None):
    if not end_time:
        end_time = time()
    return [x for x in packets if x['time_epoch'] > start_time and x['time_epoch'] < end_time]
```
This also illustrates an important python pitfall: never use function calls for default arguments:
```
>>> def test_time(t=time()):
...     print(t)
...
>>> time()
1587480434.0399244
>>> test_time()
1587480426.4636228
>>> time()
1587480438.1276047
```
See how it appears we've gone back in time? That's because the default argument is evaluated when the python interpreter first sees the function, NOT when it's called!

File: tests/test_flows.py [hunk 11]
```diff
@@ -366,10 +365,12 @@ def test_ap_capability_query(self):
...
+        
         debug("Confirming ap capability report has been received on controller")
         self.check_log(env.controller, "AP_CAPABILITY_REPORT_MESSAGE")
-
+        capability_report = env.wired_sniffer.get_packets_for_cmdu_type(self.running,0x8002)
```
Comment (ghost) [inline tests/test_flows.py]: Please remember to run `flake8` on your python scripts and fixing issues it notices. For example, here it would complain about not having a space after the comma.

File: tests/sniffer.py (hunk 2 above)
Comment (arnout) [reply tests/sniffer.py:106]: When the outputfiles dict is removed, this change is not necessary.

File: tests/sniffer.py (hunk 3 above)
Comment (arnout) [reply tests/sniffer.py]: Without -U, the output file will be buffered. That means we'd have to terminate tcpdump to see the last few packets in the output file. with the -U option, it is packet-buffered, i.e. whenever a packet is captured, the file is flushed. That way, when we read the file, we're sure that it contains everything that has been captured.

File: tests/sniffer.py [hunk 12]
```diff
@@ -7,36 +7,89 @@
...
+from time import time
 
 
 class Sniffer:
     '''Captures packets on an interface.'''
+
     def __init__(self, interface: str):
         self.interface = interface
         self.tcpdump_proc = None
+        self.outputfiles = {}
```
Comment (arnout) [inline tests/sniffer.py]: There is no need for this dictionary. At any point, there is only one running capture, thus only one running outputfile. All the get_packet_capture() calls will be for that last outputfile.

Note that multiple captures in parallel is not possible because they're all capturing the same packets.

File: tests/sniffer.py [hunk 13]
```diff
@@ -18,6 +18,7 @@ def __init__(self, interface: str):
         self.interface = interface
         self.tcpdump_proc = None
         self.outputfiles = {}
+        self.captures = {}
```
Comment (arnout) [inline tests/sniffer.py]: Again, not needed: just use the latest one.

File: tests/sniffer.py [hunk 14]
```diff
@@ -59,6 +59,20 @@ def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
         except BaseException as ex:
             return None
 
+    def _simplify_packets(self, packets):
+        def _simplify(x):
+            x['src_mac'] = x['_source']['layers']['eth']['eth.src']
```
Comment (arnout) [inline tests/sniffer.py]: (nitpick) Instead of "simplifying", it would be nicer to convert to a brand new class Ieee1905Message. Or a namedtuple instead o a full class.

File: tests/sniffer.py (hunk 8 above)
Comment (arnout) [reply tests/sniffer.py]: Yeah, there are multiple ways to go about it. Basically we want to check "was there a packet that matches these and these parameters".

I think actually the best way to do it would be to have a check_packet function in TestFlows itself that takes the filter as an argument. The filter can then check all the data it needs to.

**Review by arnout — CHANGES_REQUESTED** [review]

**Review by arnout — APPROVED** [review]
I took over this PR and fixed all the issues that were reported in this one and in #1020.

Now it is perfect :-)

**Review by tomereli — APPROVED** [review]

File: tests/sniffer.py [hunk 15]
```diff
@@ -23,7 +23,7 @@ def start(self, outputfile_basename):
         debug("Starting tcpdump, output file {}.pcap".format(outputfile_basename))
         os.makedirs(os.path.join(self.tcpdump_log_dir, 'logs'), exist_ok=True)
         outputfile = os.path.join(self.tcpdump_log_dir, outputfile_basename) + ".pcap"
-        command = ["tcpdump", "-i", self.interface, "-w", outputfile]
+        command = ["tcpdump", "-i", self.interface, '-U', '-w', outputfile]
```
Comment (rmelotte) [inline tests/sniffer.py]: Maybe it would be worth adding the `-s` option with the maximum size of a wireless frame (2346).
The default size depends on the tcpdump version, but on my machine for example it's set to 262144, while we know the maximum size we'll ever capture can't exceed 2346 (and probably even 1516 in practice).

AFAIU tcpdump sets it so high for backward compatibility, but they recommend to set it to something smaller (from man tcpdump):

> You should limit snaplen to the smallest number that will  capture the protocol information you're interested in.

(it's a nitpick comment of course, we shouldn't have that much traffic for now, so we can still do that once we start sniffing wireless frames)

**Review by rmelotte — APPROVED** [review]
I didn't review the details, overall LGTM.

I only have a nitpick comment, but it will probably only become relevant once we start sniffing wireless interfaces, so I'll add the "ready for merge" tag anyway.

File: tests/sniffer.py (hunk 15 above)
Comment (arnout) [reply tests/sniffer.py]: I believe snaplen only has an impact on memory usage, not so much on performance. So I'm not really worried about it. That said, it's a good suggestion.

BTW I intend to also drop tcpdump and replace it with tshark at some point - it's kind of stupid to require both to be installed.
